src/servers/launcher/launcher-review-request.md

```diff
@@ -0,0 +1,82 @@
+
+`launcher` caps itself off by listening for `SIGTERM` or `SIGINT`, and gracefully shutting down when such a signal arrives.
+
+I've avoided using exceptions, except in constructors, where they're used to gracefully tear down an object from within an initialiser list.  This appears to be required for RAII, but I really don't like exceptions - I don't like their ability to bubble up and kill the server.  There's little worse than playing a video game and having it _hard_ crash to the desktop because some exception deep in the bowels of a particle system that the player doesn't really care about that much sneaks by.
+
+To that end, all constructors that throw are private, and the objects are constructed with `std::expected<T, E> T::tryCreate()`.
+
+# Things I plan to do at some point and maybe don't need to be told about
+
+* There's a lambda being used to `cout` a `variant` of exceptions in `main()` which may be able to be replaced with `cout`ing the `variant` directly now I implemented `operator<<` for `variant`s.
+* There's a bug when it comes to not clearing the old packet buffer, then receiving an incomplete message.  I plan to fix this by implementing `struct Packet` that will provide `std::expected<ValidatedPacket, std::string> getValidatedPacket()`, where the `ValidatedPacket` is some view onto the underlying buffer with the transport level anti corruption checks stripped out.
+* I will be splitting this into different files.  Just wanted to get it running and learn stuff along the way, first.
+* The heartbeat will record compute/memory pressure stats in the database alongside the `launcher`'s IP in the future, such that `cluster` can pick an appropriate server on which to launch new instances.  Right now there's not a great deal of reason to update the heartbeat in the database each second as I do.
+* I've shyed away from external libraries (apart from `libpq`) mostly to learn how things work "under the hood" myself, but also in a small effort to avoid "bloat".  I'm aware Boost provides things like a UUID class and networking stuff, but I'm under the impression Boost is mega-huge, and I'd like to not use stuff like that, at least for now, unless there's an insanely compelling reason to.  (For the client, I'm sure I'll use something like raylib, so external dependencies will come).
+* String interpolation for SQL instead of paramater passing - I know.  Trust me I do.  I've gone with this way _for now_ because I want to visit [sqlc](https://sqlc.dev) for repository construction, which is going to require some work, at which point I'll deal with parameter passing, prepared statements, and probably more.
+
+# Specific things that could be better that I'd like to know about.
+
+* I haven't used RAII for saving and restoring `os.flags()` in `operator<<(std::ostream&, const UUID&)`.  It seems like it might be overkill, but it also seems like exactly the sort of place I should be using RAII.  (I'd use a context manager in Python).
```

Comment:
There is a Boost library just for this.

However, I am in the process of writing an article that deals with precisely this problem. The article is *not* finished yet, but it might still help you out. ([link](https://cpp.genesismachina.ca/article/writing-an-iostreams-inserter/))

Reply:
![image](https://github.com/user-attachments/assets/e49b3c4a-d8d2-4383-ae19-12c92dfd54fa)
I think code block styles might be broken there.  (Also can I have a dark mode please?)